libflux/src/semantic/nodes.rs

```diff
@@ -519,8 +519,74 @@ pub struct FunctionExpr {
 }
 
 impl FunctionExpr {
-    fn infer(&mut self, env: Environment, f: &mut Fresher) -> Result {
-        unimplemented!();
+    fn infer(&mut self, mut env: Environment, f: &mut Fresher) -> Result {
+        let mut cons = Constraints::empty();
+        let mut pipe = None;
+        let mut req = HashMap::new();
+        let mut opt = HashMap::new();
+        // This params will build the nested env when inferring the function body.
+        let mut params = HashMap::new();
+        for param in &mut self.params {
+            match param.default {
+                Some(ref mut e) => {
+                    let (nenv, ncons) = e.infer(env, f)?;
+                    cons = cons + ncons;
+                    let id = param.key.name.clone();
+                    // We are here: `f = (a=1) => {...}`.
+                    // So, this PolyType is actually a MonoType, whose type
+                    // is the one of the default value ("1" in "a=1").
+                    let typ = PolyType {
+                        vars: Vec::new(),
+                        cons: HashMap::new(),
+                        expr: e.type_of().clone(),
+                    };
+                    params.insert(id.clone(), typ);
+                    opt.insert(id, e.type_of().clone());
+                    env = nenv;
+                }
+                None => {
+                    // We are here: `f = (a) => {...}`.
+                    // So, we do not know the type of "a". Let's use a fresh TVar.
+                    let id = param.key.name.clone();
+                    let ftvar = f.fresh();
+                    let typ = PolyType {
+                        vars: Vec::new(),
+                        cons: HashMap::new(),
+                        expr: MonoType::Var(ftvar.clone()),
```

Comment:
Just a note, type variables are `Copy` so no need to explicitly clone. However cloning is in no way expensive for type variables.